File: CTCI/Ch2LinkedLists/Py/palindrome.py
class Node(object):
    def __init__(self, data=None, next_node=None):
        self.data = data
        self.next_node = next_node


def reverse(l):
    return l[::-1]


def LinkedList(object):
    def __init__(self, head=None):
        self.head = head

    @property
    def print_list(self):
        iterator = self.head

        while iterator is not None:
            print('{}-->'.format(iterator.data))
            iterator = iterator.next_node

    def insert(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        new_node.next_node = self.head
        self.head = new_node

    @property
    def palindrome(self):
        iterator = self.head
        tl = []
        while iterator is not None:
            tl.append(iterator.data)
            iterator = iterator.next

        compare = reverse(tl)
        return tl == compare


def main():
    ll = LinkedList(None)
    d = [letter for letter in 'racecar']
    print(d)
    for letter in d:
        ll.insert(d)

    print(ll.palindrome)


main()

File: CTCI/Ch2LinkedLists/java/Palindrome.java
import java.util.*;
/**
  * This code checks to see if a Linked List
  * of characters contains a palindrome 
  * (a word that is the same front to back)
  * Example:
  * m-->o-->m true
  * m-->o-->k false
  */

public class Palindrome {

	protected Node head;
	
	public class Node {
		char data;
		Node next;

		public Node() {
			next = null;
		}

		public Node(char newData) {
			data = newData;
		}
	}

	public void insert(char data) {
		if (head == null) {
			head = new Node(data);
		} else {
			Node newVal = new Node(data);
			newVal.next = head;
			head = newVal;
		}
	}

	public void print(Node head) {
		Node temp = head;
		while (temp != null) {
			System.out.print(temp.data + "-->");
			temp = temp.next;
		}
	}

	public boolean isPalindrome(Node root) {
		Node temp = root;
		HashMap<Character, Integer> charMap = new HashMap<>();
		while (temp != null) {
			int count = charMap.containsKey(temp.data) ? charMap.get(temp.data) : 0;
			charMap.put(temp.data, count + 1);
			temp = temp.next;
		}

		int differences = 0;
		for (int vals : charMap.values()) {
			if (vals %2 == 0) {
				continue;
			} else {
				differences++;
			}
		}
		return differences <= 1;
	}

	public static void main(String[] args) {
		Palindrome p = new Palindrome();
		p.insert('m');
		p.insert('o');
		p.insert('k');
		System.out.println(p.isPalindrome(p.head));
	}
}
